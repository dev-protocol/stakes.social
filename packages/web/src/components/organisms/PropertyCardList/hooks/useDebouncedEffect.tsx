/* eslint-disable react-hooks/exhaustive-deps */
import { useEffect } from 'react'

const useDebouncedEffect = (callback: () => any, delay = 250, dependencies?: any[]) =>
  useEffect(() => {
    const timer = setTimeout(callback, delay)

    return () => {
      clearTimeout(timer)
    }
  }, dependencies)

export default useDebouncedEffect
