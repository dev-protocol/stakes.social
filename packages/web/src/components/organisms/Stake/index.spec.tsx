import React from 'react'
import Web3 from 'web3'
import BigNumber from 'bignumber.js'
import { message } from 'antd'
import { act, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Stake } from '.'
import { balanceOf } from 'src/fixtures/dev-kit/client'
import { useProvider } from 'src/fixtures/wallet/hooks'
import { whenDefined } from 'src/fixtures/utility'
import 'src/__mocks__/window/matchMedia.mock'
import WalletContext from 'src/context/walletContext'
import { getContractAddress } from 'src/fixtures/dev-kit/get-contract-address'

jest.mock('src/fixtures/dev-kit/client')
jest.mock('src/fixtures/dev-kit/hooks')
jest.mock('src/fixtures/dev-kit/get-contract-address')
jest.mock('src/fixtures/uniswap/hooks')
jest.mock('src/fixtures/wallet/hooks')
jest.mock('web3')
jest.mock('@dev/graphql')

describe(`${Stake.name}`, () => {
  test('Snapshot', () => {
    const component = render(
      <WalletContext.Provider value={{ web3: new Web3(), setProviders: () => {} }}>
        <Stake propertyAddress="propertyAddress" />
      </WalletContext.Provider>
    )
    const tree = component.baseElement
    expect(tree).toMatchSnapshot()
  })

  test('Snapshot with title', () => {
    const component = render(
      <WalletContext.Provider value={{ web3: new Web3(), setProviders: () => {} }}>
        <Stake propertyAddress="propertyAddress" title="Stake" />
      </WalletContext.Provider>
    )
    const tree = component.baseElement
    expect(tree).toMatchSnapshot()
  })

  test('react hooks: input amount click DEV button and click stake button', async () => {
    ;(balanceOf as jest.Mock).mockImplementation(() => Promise.resolve(new BigNumber('1000000')))
    ;(useProvider as jest.Mock).mockImplementation(() => ({ accountAddress: '', ethersProvider: {} }))
    ;(getContractAddress as jest.Mock).mockImplementation(() => Promise.resolve('0x0'))
    message.warn = jest.fn(() => {}) as any

    const { container, getByText } = render(
      <WalletContext.Provider value={{ web3: new Web3(), setProviders: () => {} }}>
        <Stake propertyAddress="propertyAddress" />
      </WalletContext.Provider>
    )

    const input = container.querySelector('input#stake')
    expect(input).not.toBe(null)
    await act(async () => {
      await whenDefined(input, x => userEvent.type(x, '12'))
      await userEvent.click(getByText('DEV'))
      await userEvent.click(getByText('Approve'))
    })

    expect((message.warn as jest.Mock).mock.calls.length).toBe(0)
  })

  // TODO(@k3nt0w)
  // test('react hooks: click DEV button(amount is zero) and click stake button', async () => {
  //   ;(balanceOf as jest.Mock).mockImplementation(() => Promise.resolve(new BigNumber('')))
  //   ;(useProvider as jest.Mock).mockImplementation(() => ({ accountAddress: '', ethersProvider: {} }))
  //   message.warn = jest.fn(() => {}) as any

  //   const { container, getByText } = render(
  //     <WalletContext.Provider value={{ web3: new Web3(), setProviders: () => {} }}>
  //       <Stake propertyAddress="propertyAddress" />
  //     </WalletContext.Provider>
  //   )

  //   const input = container.querySelector('input#stake')
  //   expect(input).not.toBe(null)
  //   await act(async () => {
  //     await userEvent.click(getByText('DEV'))
  //     await userEvent.click(getByText('Stake'))
  //   })

  //   expect((message.warn as jest.Mock).mock.calls.length).toBe(1)
  // })
})
