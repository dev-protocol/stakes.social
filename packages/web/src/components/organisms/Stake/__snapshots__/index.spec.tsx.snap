// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stake Snapshot 1`] = `
<body>
  <div>
    <div
      class="sc-1edrou6-0 ehtZMB"
    >
      <div
        class="sc-1nbnut-4 kmrGix"
        style="opacity: 1;"
      >
        <div
          class="ant-radio-group ant-radio-group-outline"
          style="margin-bottom: 12px;"
        >
          <div
            class="ant-space ant-space-vertical"
          >
            <div
              class="ant-space-item"
              style="margin-bottom: 8px;"
            >
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="-1"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  <span
                    style="margin-left: 12px;"
                  >
                    New position
                  </span>
                </span>
              </label>
            </div>
            <div
              class="ant-space-item"
              style="margin-bottom: 8px;"
            >
              )
            </div>
            <div
              class="ant-space-item"
              style="margin-bottom: 8px;"
            >
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="2"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="margin-left: 8px; min-width: 40px;"
                      >
                        #
                        2
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="min-width: 120px;"
                      >
                        <span>
                          0.000000000000001
                        </span>
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Staked
                        </span>
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                    >
                      <div>
                        0
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Claimable
                        </span>
                      </div>
                    </div>
                  </div>
                </span>
              </label>
            </div>
            <div
              class="ant-space-item"
              style="margin-bottom: 8px;"
            >
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="4"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="margin-left: 8px; min-width: 40px;"
                      >
                        #
                        4
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="min-width: 120px;"
                      >
                        <span>
                          0.000000000000001
                        </span>
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Staked
                        </span>
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                    >
                      <div>
                        0
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Claimable
                        </span>
                      </div>
                    </div>
                  </div>
                </span>
              </label>
            </div>
            <div
              class="ant-space-item"
            >
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="6"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="margin-left: 8px; min-width: 40px;"
                      >
                        #
                        6
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="min-width: 120px;"
                      >
                        <span>
                          0.000000000000001
                        </span>
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Staked
                        </span>
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                    >
                      <div>
                        0
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Claimable
                        </span>
                      </div>
                    </div>
                  </div>
                </span>
              </label>
            </div>
          </div>
        </div>
        <span
          class="ant-input-affix-wrapper ant-input-affix-wrapper-lg sc-1nbnut-0 fRofMr"
        >
          <input
            class="ant-input ant-input-lg"
            id="stake"
            type="number"
            value=""
          />
          <span
            class="ant-input-suffix"
          >
            <button
              class="sc-1ttu0x8-0 kmiVBa"
            >
              DEV
            </button>
          </span>
        </span>
        <div
          class="sc-1nbnut-2 ihkVKZ"
        >
          <div
            class="ant-row"
          >
            <div
              class="ant-col ant-col-11"
            >
              <button
                class="ant-btn ant-btn-primary sc-1nbnut-1 gUxzv"
                type="button"
              >
                <span>
                  Approve
                </span>
              </button>
            </div>
            <div
              class="ant-col ant-col-2 sc-1nbnut-3 kdZRmt"
            >
              <svg
                fill="none"
                height="24"
                transform="rotate(180)"
                viewBox="0 0 32 24"
                width="32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M32 13.7611V10.2389L6.5256 10.2389L6.5256 8.60068C8.49147 8.35495 9.63823 8.0273 11.4812 6.26621L15.3311 2.45734L12.8328 0L0.873718 12L12.8328 24L15.3311 21.5427L11.2765 17.529C9.76109 16.0546 8.53242 15.686 6.5256 15.3993V13.7611L32 13.7611Z"
                  fill="#5B8BF5"
                />
              </svg>
            </div>
            <div
              class="ant-col ant-col-11"
            >
              <button
                class="ant-btn ant-btn-primary sc-1nbnut-1 gUxzv"
                disabled=""
                type="button"
              >
                <span>
                  Stake
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        style="height: 40px;"
      />
      <div
        class="ant-card ant-card-bordered ant-card-small sc-18v7ikh-0 dfjrvZ"
      >
        <div
          class="ant-card-head"
        >
          <div
            class="ant-card-head-wrapper"
          >
            <div
              class="ant-card-head-title"
            >
              Gas Fee (predicted)
            </div>
          </div>
        </div>
        <div
          class="ant-card-body"
        >
          <p>
            0.003345
             ETH
            <span
              class="sc-18v7ikh-1 kGdgpU"
            >
               $0.78
            </span>
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Stake Snapshot with title 1`] = `
<body>
  <div>
    <div
      class="sc-1edrou6-0 ehtZMB"
    >
      <label
        for="stake"
      >
        Stake
      </label>
      <div
        class="sc-1nbnut-4 kmrGix"
        style="opacity: 1;"
      >
        <div
          class="ant-radio-group ant-radio-group-outline"
          style="margin-bottom: 12px;"
        >
          <div
            class="ant-space ant-space-vertical"
          >
            <div
              class="ant-space-item"
              style="margin-bottom: 8px;"
            >
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="-1"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  <span
                    style="margin-left: 12px;"
                  >
                    New position
                  </span>
                </span>
              </label>
            </div>
            <div
              class="ant-space-item"
              style="margin-bottom: 8px;"
            >
              )
            </div>
            <div
              class="ant-space-item"
              style="margin-bottom: 8px;"
            >
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="2"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="margin-left: 8px; min-width: 40px;"
                      >
                        #
                        2
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="min-width: 120px;"
                      >
                        <span>
                          0.000000000000001
                        </span>
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Staked
                        </span>
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                    >
                      <div>
                        0
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Claimable
                        </span>
                      </div>
                    </div>
                  </div>
                </span>
              </label>
            </div>
            <div
              class="ant-space-item"
              style="margin-bottom: 8px;"
            >
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="4"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="margin-left: 8px; min-width: 40px;"
                      >
                        #
                        4
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="min-width: 120px;"
                      >
                        <span>
                          0.000000000000001
                        </span>
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Staked
                        </span>
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                    >
                      <div>
                        0
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Claimable
                        </span>
                      </div>
                    </div>
                  </div>
                </span>
              </label>
            </div>
            <div
              class="ant-space-item"
            >
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="6"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="margin-left: 8px; min-width: 40px;"
                      >
                        #
                        6
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                      style="margin-right: 24px;"
                    >
                      <div
                        style="min-width: 120px;"
                      >
                        <span>
                          0.000000000000001
                        </span>
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Staked
                        </span>
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                    >
                      <div>
                        0
                        <span
                          style="font-size: 0.6em; margin-left: 2px;"
                        >
                          Claimable
                        </span>
                      </div>
                    </div>
                  </div>
                </span>
              </label>
            </div>
          </div>
        </div>
        <span
          class="ant-input-affix-wrapper ant-input-affix-wrapper-lg sc-1nbnut-0 fRofMr"
        >
          <input
            class="ant-input ant-input-lg"
            id="stake"
            type="number"
            value=""
          />
          <span
            class="ant-input-suffix"
          >
            <button
              class="sc-1ttu0x8-0 kmiVBa"
            >
              DEV
            </button>
          </span>
        </span>
        <div
          class="sc-1nbnut-2 ihkVKZ"
        >
          <div
            class="ant-row"
          >
            <div
              class="ant-col ant-col-11"
            >
              <button
                class="ant-btn ant-btn-primary sc-1nbnut-1 gUxzv"
                type="button"
              >
                <span>
                  Approve
                </span>
              </button>
            </div>
            <div
              class="ant-col ant-col-2 sc-1nbnut-3 kdZRmt"
            >
              <svg
                fill="none"
                height="24"
                transform="rotate(180)"
                viewBox="0 0 32 24"
                width="32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M32 13.7611V10.2389L6.5256 10.2389L6.5256 8.60068C8.49147 8.35495 9.63823 8.0273 11.4812 6.26621L15.3311 2.45734L12.8328 0L0.873718 12L12.8328 24L15.3311 21.5427L11.2765 17.529C9.76109 16.0546 8.53242 15.686 6.5256 15.3993V13.7611L32 13.7611Z"
                  fill="#5B8BF5"
                />
              </svg>
            </div>
            <div
              class="ant-col ant-col-11"
            >
              <button
                class="ant-btn ant-btn-primary sc-1nbnut-1 gUxzv"
                disabled=""
                type="button"
              >
                <span>
                  Stake
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        style="height: 40px;"
      />
      <div
        class="ant-card ant-card-bordered ant-card-small sc-18v7ikh-0 dfjrvZ"
      >
        <div
          class="ant-card-head"
        >
          <div
            class="ant-card-head-wrapper"
          >
            <div
              class="ant-card-head-title"
            >
              Gas Fee (predicted)
            </div>
          </div>
        </div>
        <div
          class="ant-card-body"
        >
          <p>
            0.003345
             ETH
            <span
              class="sc-18v7ikh-1 kGdgpU"
            >
               $0.78
            </span>
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
